### PR TITLE
feat: pagninate status poller lists

### PR DESCRIPTION
**Pull request by karlkfi**
- Use the client-go ListPager to paginate requests made by the
  CachingClusterReader used by the StatusPoller.
- Limit requests to 500 items per response (ListPager default)
- Benefits:
  - Reduce the length of blocking etcd operations, allowing for more
    responsiveness and higher throughput
  - Reduce risk of request timeout
  - Reduce size of response for extra large objects

### CODE REVIEW

File: pkg/kstatus/polling/clusterreader/caching_reader.go [hunk 1]
```diff
@@ -254,9 +257,82 @@ func (c *CachingClusterReader) Sync(ctx context.Context) error {
...
+) (*unstructured.UnstructuredList, error) {
+	mOpts := metav1.ListOptions{}
+	mOpts.SetGroupVersionKind(gvk)
+	obj, _, err := pager.New(c.listPageFunc(namespace)).List(ctx, mOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	switch t := obj.(type) {
+	case *unstructured.UnstructuredList:
```
Comment (mortent) [inline pkg/kstatus/polling/clusterreader/caching_reader.go:283]: Can we add a comment here to explain why we could get the two different return types here?

**Review by mortent — APPROVED** [review]
/approve

**Comment by karlkfi** [comment]
With Pagination:
```
• [SLOW TEST:897.772 seconds]
Applier
/home/karlisenberg/workspace/cli-utils/test/stress/stress_test.go:36
  StressTest
  /home/karlisenberg/workspace/cli-utils/test/stress/stress_test.go:77
    Thousand Namespaces
    /home/karlisenberg/workspace/cli-utils/test/stress/stress_test.go:98
------------------------------

Ran 1 of 1 Specs in 897.860 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 15m2.181886894s
Test Suite Passed
```

Without Pagination:
```
• [SLOW TEST:848.551 seconds]
Applier
/home/karlisenberg/workspace/cli-utils/test/stress/stress_test.go:36
  StressTest
  /home/karlisenberg/workspace/cli-utils/test/stress/stress_test.go:77
    Thousand Namespaces
    /home/karlisenberg/workspace/cli-utils/test/stress/stress_test.go:98
------------------------------

Ran 1 of 1 Specs in 848.618 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 14m13.658096361s
Test Suite Passed
```

7% time diff for the whole test may or may not be statistically significant. Could easily have just been another trip through the StatusPoller checking 3k objects. :shrug:

**Comment by Liujingfang1** [comment]
Thanks for the timing data.

/lgtm

**Pull request merged**